tianshou/policy/modelbase/psrl.py

```diff
@@ -0,0 +1,197 @@
+import numpy as np
+from typing import Any, Dict, Union, Optional
+
+from tianshou.policy import BasePolicy
+from tianshou.data import Batch, ReplayBuffer
+
+
+class PSRLModel(object):
+    """Implementation of Posterior Sampling Reinforcement Learning Model.
+
+    :param np.ndarray p_prior: dirichlet prior (alphas), with shape
+        (n_action, n_state, n_state).
+    :param np.ndarray rew_mean_prior: means of the normal priors of rewards,
+        with shape (n_state, n_action).
+    :param np.ndarray rew_std_prior: standard deviations of the normal priors
+        of rewards, with shape (n_state, n_action).
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.BasePolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        p_prior: np.ndarray,
+        rew_mean_prior: np.ndarray,
+        rew_std_prior: np.ndarray,
+    ) -> None:
+        self.__eps = np.finfo(np.float32).eps.item()
+        self.p = p_prior
+        self.n_action, self.n_state, _ = p_prior.shape
+        self.rew_mean = rew_mean_prior
+        self.rew_std = rew_std_prior
+        self.rew_count = np.zeros_like(rew_mean_prior) + self.__eps
+        self.policy: Optional[np.ndarray] = None
+        self.updated = False
+
+    def observe(
+        self, p: np.ndarray, rew_sum: np.ndarray, rew_count: np.ndarray
+    ) -> None:
+        """Add data into memory pool.
+
+        For rewards, we have a normal prior at first. After we observed a
+        reward for a given state-action pair, we use the mean value of our
+        observations instead of the prior mean as the posterior mean. The
+        standard deviations are in inverse proportion to the number of the
+        corresponding observations.
+
+        :param np.ndarray p: the number of observations, with shape
+            (n_action, n_state, n_state).
+        :param np.ndarray rew_sum: total rewards, with shape
+            (n_state, n_action).
+        :param np.ndarray rew_count: the number of rewards, with shape
+            (n_state, n_action).
+        """
+        self.updated = False
+        self.p += p
+        sum_count = self.rew_count + rew_count
+        self.rew_mean = (self.rew_mean * self.rew_count + rew_sum) / sum_count
+        self.rew_std *= self.rew_count / sum_count
+        self.rew_count = sum_count
+
+    def get_p_ml(self) -> np.ndarray:
+        return self.p / np.sum(self.p, axis=-1, keepdims=True)
+
+    def sample_from_rew(self) -> np.ndarray:
+        sample_rew = np.random.randn(*self.rew_mean.shape)
+        sample_rew = sample_rew * self.rew_std + self.rew_mean
+        return sample_rew
+
+    def solve_policy(self) -> None:
+        self.updated = True
+        self.policy = self.value_iteration(
+            self.get_p_ml(), self.sample_from_rew())
+
+    @staticmethod
+    def value_iteration(
+        p: np.ndarray, rew: np.ndarray, eps: float = 0.01
```

Comment:
this "p" is transition prob, so it should be named as "trans_prob" to distinguish it with the "trans_count"